### PR TITLE
[GOBBLIN-822] Upgrade log4j to log4j2 for LogRouting use-case

### PR DESCRIPTION
**Pull request by jhsenjaliya**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-822

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
log4j2 has routing appender that would be super useful and probably only way to achieve "job specific logging" functionality without meddling around fileHandler in log4j.
Also log4j2 has lot of new functionalities and performance benefits
==NOTE== there are PRs that is not merged into master, hence we see some commit messages and changes from previous commits, that i have rebased to, once that PR is merged, those changes will go away from this PR.==

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason: No new functionality added

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by jhsenjaliya** [comment]
cant find any specific error in the build, so please review this and i will kick of another build later. Thanks

File: gobblin-test-harness/src/test/java/org/apache/gobblin/TaskErrorIntegrationTest.java [hunk 1]
```diff
@@ -114,24 +113,24 @@ public void testErrorDuringSubmission()
...
 
     logger.removeAppender(testAppender);
   }
 
   @Test
   public void testCustomizedTaskFrameworkFailureInTaskCreation() throws Exception {
-    TestAppender testAppender = new TestAppender();
-    Logger logger = LogManager.getLogger(GobblinMultiTaskAttempt.class.getName() + "-noattempt");
+    TestAppender testAppender = new TestAppender(GobblinMultiTaskAttempt.class.getName() + "-noattempt");
+    Logger logger = (Logger) LogManager.getLogger(GobblinMultiTaskAttempt.class.getName() + "-noattempt");
```
Comment (arjun4084346) [inline gobblin-test-harness/src/test/java/org/apache/gobblin/TaskErrorIntegrationTest.java]: Maybe change to 
Logger logger = (Logger) LogManager.getLogger(testAppender.getName()); ?

also in all the places above

File: gobblin-test-harness/src/test/java/org/apache/gobblin/TaskErrorIntegrationTest.java [hunk 2]
```diff
@@ -232,12 +231,16 @@ public void shutdown(SourceState state) {
     }
   }
 
-  private class TestAppender extends AppenderSkeleton {
-    List<LoggingEvent> events = new ArrayList<LoggingEvent>();
-    public void close() {}
-    public boolean requiresLayout() {return false;}
+  public class TestAppender extends AbstractAppender {
```
Comment (arjun4084346) [inline gobblin-test-harness/src/test/java/org/apache/gobblin/TaskErrorIntegrationTest.java]: Can we use one class of TestAppender instead of creating one in different test classes?
Definitely need not be the part of this PR.

**Comment by arjun4084346** [comment]
LGTM. Thanks for the improvements.

**Comment by arjun4084346** [comment]
Just a minor comment. Also one test case is failing, please make sure travis run passes.

File: gobblin-test-harness/src/test/java/org/apache/gobblin/TaskErrorIntegrationTest.java (hunk 2 above)
Comment (jhsenjaliya) [reply gobblin-test-harness/src/test/java/org/apache/gobblin/TaskErrorIntegrationTest.java]: ya, since it was getting big, wanted to avoid more refactor, I will create minor ticket for this.

File: gobblin-test-harness/src/test/java/org/apache/gobblin/TaskErrorIntegrationTest.java (hunk 1 above)
Comment (jhsenjaliya) [reply gobblin-test-harness/src/test/java/org/apache/gobblin/TaskErrorIntegrationTest.java]: i tired changing the name but the test case kept failing, even though there should not be any reason to fail, but not sure what is going on, I will check on that later, for now, I will have to keep the same.

**Comment by jhsenjaliya** [comment]
@arjun4084346 , Please review.

**Comment by jhsenjaliya** [comment]
@arjun4084346 , @sv2000 , please take a look at this, this should be ready to get merged now. Thanks

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/version/finder/AbstractHiveDatasetVersionFinder.java [hunk 3]
```diff
@@ -68,7 +68,7 @@
     }
     final HiveDataset hiveDataset = (HiveDataset) dataset;
 
-    if (!HiveUtils.isPartitioned(hiveDataset.getTable())) {
+    if (!hiveDataset.getTable().isPartitioned()) {
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/version/finder/AbstractHiveDatasetVersionFinder.java]: Same comment.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveUtils.java [hunk 4]
```diff
@@ -60,7 +60,7 @@
    * @param table the {@link Table} for which we should get partitions.
    * @param filter an optional filter for partitions as would be used in Hive. Can only filter on String columns.
    *               (e.g. "part = \"part1\"" or "date > \"2015\"".
-   * @return a map of values to {@link Partition} for input {@link Table}.
+   * @return a map of values to {@link Partition} for input {@link Table} filtered and non-nullified.
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveUtils.java]: Seems unrelated to log4j->log4j2.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveUtils.java [hunk 5]
```diff
@@ -165,6 +165,7 @@ private static Configuration getHadoopConfiguration() {
 
   /**
    * @return true if {@link Table} is partitioned.
+   * @deprecated use {@link Table}'s isPartitioned method directly.
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveUtils.java]: Same comment as above.

File: gobblin-docs/case-studies/Hive-Distcp.md [hunk 6]
```diff
@@ -11,7 +11,7 @@ This document will show an sample job config of running Gobblin hive distcp, and
 
 # Configure Hive Distcp Job
 
-Below is the sample job config of running Gobblin hive distcp. Gobblin job constructs and data flow are the same as [Gobblin distcp](http://gobblin.readthedocs.io/en/latest/adaptors/Gobblin-Distcp/). The only difference is the `gobblin.data.profile.class` and hive related properties.
+Below is the sample job config of running Gobblin hive distcp. Gobblin job constructs and data flow are the same as [Gobblin distcp](http://gobblin.readthedocs.io/en/latest/adaptors/Gobblin-Distcp/). The only difference is the `gobblin.dataset.profile.class` and hive related properties.
```
Comment (sv2000) [inline gobblin-docs/case-studies/Hive-Distcp.md]: Same comment as above.

File: gobblin-docs/user-guide/Configuration-Properties-Glossary.md [hunk 7]
```diff
@@ -250,7 +250,7 @@ These properties are common properties that are used among different Source impl
 ### DistcpFileSplitter Properties <a name="DistcpFileSplitter-Properties"></a>
 | Name | Description | Required | Default Value |
 | --- | --- | --- | --- |
-| `gobblin.copy.split.enabled` | Will split files into block level granularity work units, which can be copied independently, then merged back together before publishing. To actually achieve splitting, the max split size property also needs to be set. | No | False| 
+| `gobblin.copy.split.enabled` | Will split files into block level granularity work units, which can be copied independently, then merged back together before publishing. To actually achieve splitting, the max split size property also needs to be set. Splitting should not be used for distcp jobs that require decryption/ungzipping. | No | False|
```
Comment (sv2000) [inline gobblin-docs/user-guide/Configuration-Properties-Glossary.md]: Same comment as above.

File: gobblin-modules/gobblin-azkaban/src/main/java/org/apache/gobblin/data/management/conversion/hive/validation/ValidationJob.java [hunk 8]
```diff
@@ -293,7 +293,7 @@ private void runCountValidation()
 
           // Validate dataset
           log.info(String.format("Validating dataset: %s", hiveDataset));
-          if (HiveUtils.isPartitioned(hiveDataset.getTable())) {
+          if (hiveDataset.getTable().isPartitioned()) {
```
Comment (sv2000) [inline gobblin-modules/gobblin-azkaban/src/main/java/org/apache/gobblin/data/management/conversion/hive/validation/ValidationJob.java]: Same comment as above.

**Comment by jhsenjaliya** [comment]
@sv2000, This PR was created on top of GOBBLIN-785, so can u pls you merge #2650 first ? Thanks

**Comment by sv2000** [comment]
@jhsenjaliya Since PR 2650 is merged can you clean up this PR?

**Comment by jhsenjaliya** [comment]
Done, rebased it with GOBBLIN-785, Thanks

**Comment by Will-Lo** [comment]
I'm running the scripts from `bin`,  and it appears these changes cause many of the services to stop outputting logs. @jhsenjaliya would you know what's causing this?
